[DO NOT MERGE] refactor: attempt to reduce rerenders in nft grid

Prithpal-Sooriya opened the pull request:
## **Description**

DO NOT MERGE.

Attempting to optimise the performance/rerenders inside the nftgrid list. As it seems to crash in dev due to unoptimised renders.

> VirtualizedList: You have a large list that is slow to update - make sure your renderItem function renders components that follow React performance best practices like PureComponent, shouldComponentUpdate, etc. {"contentLength": 2984.727294921875, "dt": 24841, "prevDt": 1225}


## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

Prithpal-Sooriya committed:
refactor: attempt to reduce rerenders in nft grid

Prithpal-Sooriya commented on app/components/UI/NftGrid/NftGrid.tsx:
```diff
@@ -63,8 +69,13 @@ function NftGrid({ chainId, selectedAddress }: NftGridProps) {
     >();
-  const collectibles = useSelector(collectiblesSelector).filter(
-    (singleCollectible: Nft) => singleCollectible.isCurrentlyOwned === true,
+  const allCollectibles = useSelector(collectiblesSelector);
+  const collectibles = useMemo(
```
NOTE - before .filter would return a new array, so _potentially_ was causing new renders.

Changed to useMemo + filter to reduce this, but unsure if this actually made a tangible impact.

Prithpal-Sooriya commented on app/components/UI/NftGrid/NftGrid.tsx:
```diff
@@ -222,6 +233,11 @@ function NftGrid({ chainId, selectedAddress }: NftGridProps) {
     useNftDetection,
   ]);
 
+  const refreshColors = useMemo(
+    () => [colors.primary.default],
+    [colors.primary.default],
+  );
```
Easy win, the footer was re-rendering since this was a new array on each render cycle.

But not the main issue...

Prithpal-Sooriya commented on app/components/UI/NftGrid/NftGrid.tsx:
```diff
@@ -241,7 +257,7 @@ function NftGrid({ chainId, selectedAddress }: NftGridProps) {
       )}
       {/* nft grid */}
-      {!isNftFetchingProgress && collectibles.length > 0 && (
+      {collectibles.length > 0 && (
```
FIXME - this is a temp change to get this list rendered.

sahar-fehri replied:
It does seem like because the state of nfts keeps growing, the component keeps rendering resulting eventually in that error we saw; i wonder if there is a react trick to fix this, maybe passing collectibles as a param to the component? 🤔

Prithpal-Sooriya commented on app/components/UI/NftGrid/NftGridItem.tsx:
```diff
@@ -38,29 +38,22 @@ function NftGridItem({
-    },
-    [navigation],
-  );
+  const onItemPress = useCallback(() => {
+    debouncedNavigation(navigation, nft);
+  }, [navigation, nft]);
```
nit - was a prettier change

Prithpal-Sooriya commented on app/components/UI/NftGrid/NftGridItem.tsx:
```diff
@@ -38,29 +38,22 @@ function NftGridItem({
       style={styles.collectibleCard}
-      onPress={() => onItemPress(nft)}
-      onLongPress={() => onLongPressCollectible(nft)}
+      onPress={onItemPress}
+      onLongPress={onLongPressCollectible}
```
nit - changed the function interface so that we don't re-create these functions on renders.

Prithpal-Sooriya commented on app/components/UI/NftGrid/NftGridItem.tsx:
```diff
@@ -79,4 +72,4 @@ function NftGridItem({
 }
 
-export default NftGridItem;
+export default React.memo(NftGridItem);
```
Made this component memoized. However I'm unsure how much impact this made.

Prithpal-Sooriya commented on app/components/UI/NftGrid/NftGridItem.tsx:
```diff
@@ -38,29 +38,22 @@ function NftGridItem({
 
   return (
     <TouchableOpacity
-      key={nft.address}
```
Removed the `key` field as we already use the FlatList.keyExtractor to generate the key.

Prithpal-Sooriya committed:
perf: try reducing some re-renders

Prithpal-Sooriya commented on app/components/Base/RemoteImage/index.js:
```diff
@@ -199,7 +199,7 @@ const RemoteImage = (props) => {
                   badgeElement={
                     <Badge
                       variant={BadgeVariant.Network}
-                      imageSource={NetworkBadgeSource()}
```
Nit - react dev tools pointed at this re-rendering, but it probably is miniscule.

Prithpal-Sooriya commented on app/components/UI/NftGrid/NftGrid.tsx:
```diff
@@ -58,13 +64,75 @@ interface NftGridProps {
+  return nfts;
+}
+
+export const useNfts = () => {
```
Shim as using the real API will end up with http 429 rate limit.

This shim will add 5 NFTs per 500ms. Meh we can tweak it, and not sure if realistic.

Prithpal-Sooriya commented on app/components/UI/NftGrid/NftGrid.tsx:
```diff
@@ -222,6 +290,29 @@ function NftGrid({ chainId, selectedAddress }: NftGridProps) {
+    [navigation, privacyMode],
+  );
+
+  const keyExtractor = useCallback(
```
React dev tools pointed at these also causing prop changes.

Prithpal-Sooriya commented on app/components/UI/NftGrid/NftGridFooter.tsx:
```diff
@@ -17,22 +17,26 @@ interface NftGridFooterProps {
   return (
-    <View
-      // eslint-disable-next-line react-native/no-inline-styles
-      style={{
```
React dev tools pointed at this footer re-rendering everytime the list changes. It was due to this unstable prop.

However this footer is pretty small compared to the list items.

Prithpal-Sooriya commented on app/components/UI/NftGrid/NftGridFooter.tsx:
```diff
@@ -17,22 +17,26 @@ interface NftGridFooterProps {
         {strings('wallet.no_collectibles')}
       </Text>
       <TouchableOpacity
-        onPress={() =>
```
likewise, this is an unstable prop.

Prithpal-Sooriya commented on app/reducers/notification/index.js:
```diff
@@ -26,8 +26,10 @@ const enqueue = (notifications, notification) => [
+
 export const currentNotificationSelector = (state) =>
-  state?.notifications[0] || {};
+  state?.notifications[0] || emptyObj;
```
Ignore this - it was polluting the react dev tools as it was re-rendering almost on every render cycle. I might separate this as a small fix.

Prithpal-Sooriya committed:
perf: throttle selector data so we don't continuously render

IDK if throttle is the right approach, since it might lose the last call? Maybe we can retain this last call so it isn't lost?

But this does help reduce number of renders since this selector fires a lot (since it continuously fetches)

Prithpal-Sooriya commented on app/components/UI/NftGrid/NftGrid.tsx:
```diff
@@ -58,13 +65,80 @@ interface NftGridProps {
+  return nfts;
+};
+
+const throttleFilterOwnedCollectibles = throttle(
+  (collectibles: Nft[]) =>
+    collectibles.filter(
+      (singleCollectible: Nft) => singleCollectible.isCurrentlyOwned === true,
+    ),
+  1000,
+);
+
```
Using a throttle on the memoized filter, mostly because the selector is recalled many times (since we are continuously fetching data).

So at least now the collectibles are memoized every second (so the list only renders every second).

Meh, maybe this is good... IDK. Worried that a throttle will prevent rendering of the last page of data. Maybe there is a leading/trailing/flush options to use here?

Prithpal-Sooriya committed:
perf: remove additional SVG fetching when determining size

Prithpal-Sooriya commented:
Closing as this is not necessary anymore
